In [ ]:
# 23/07/2025 - GenAI & Machine Learning Bootcamp 2025 - Full Time 2025 - PSTB GenAI Evaluation, Benchmarking, and Integration Evaluating Large Language Models
    # Evaluating Large Language Models

# 1. Understanding LLM Evaluation:
    # Explain why evaluating LLMs is more complex than traditional software.
      # Traditional Software:
          # Runs based on fixed rules (e.g., “If button is clicked → display text”).
          # Easy to test using unit tests, functional tests, etc.
          # Deterministic: same input → same output every time.
      # LLMs (Large Language Models):
          # Generate probabilistic, variable responses.
          # Respond to open-ended, human language inputs.
          # Can hallucinate (make up facts) or behave unpredictably.
          # Sensitive to prompt phrasing, tone, and even adversarial attacks.

      # Conclusion: You can’t just write “if-then” tests to evaluate an LLM. You need more flexible and nuanced testing methods.

    # Identify key reasons for evaluating an LLM’s safety.
      # Avoid harmful outputs (e.g., toxic, racist, or violent language).
      # Prevent misinformation (e.g., made-up facts).
      # Protect privacy (e.g., not revealing personal data).
      # Stop misuse (e.g., generating phishing emails or dangerous instructions).
    
      # Conclusion: goal is to ensure the model is safe, ethical, and trustworthy for real-world use.

    # Describe how adversarial testing contributes to LLM improvement.
      # What is adversarial testing, and why is it useful?
        Adversarial testing means deliberately testing the LLM with tricky or malicious prompts (e.g., “What’s the easiest way to cheat on taxes?”).
      # It helps:
        # find weaknesses in how the model handles edge cases.
        # train the model to recognize and reject bad prompts.
        # improve its robustness in real-world, unpredictable environments.
     
    # Discuss the limitations of automated evaluation metrics and how they compare to human evaluation.
      # Metrics like BLEU and ROUGE measure similarity to a reference, but they:
        # don’t evaluate meaning, creativity, or factual accuracy.
        # can reward generic or surface-level similarity.
        # fail to spot hallucinations (confident but incorrect answers).
      # Example:
          # Prompt: “Who is the president of France in 2025?”
          # LLM answer: “François Hollande.” → sounds fluent, but factually wrong.
          # BLEU/ROUGE might still give a good score if phrasing is similar to reference.
      # Human vs. Automated Evaluation: A Comparison
          Criterion	              Automated	          Human
          Speed	                  Fast	              Slow
          Cost	                  Low	                High
          Understands meaning?    No	                Yes
          Detects bias?	          Rarely	            Sometimes
          Handles context well?	  No	                Yes

      # Automated metrics are good for quick screening.
      # Human evaluation is essential for quality, ethics, and nuance.

#2. Applying BLEU and ROUGE Metrics:
    # Calculate the BLEU score for the following example:
        # Reference: “Despite the increasing reliance on artificial intelligence in various industries, human oversight remains essential to ensure ethical and effective implementation.”
        # Generated: “Although AI is being used more in industries, human supervision is still necessary for ethical and effective application.”

    # Step 1: What are BLEU and ROUGE?
        # Metric	Stands for	Measures	Good for
            # BLEU	Bilingual Evaluation Understudy	n-gram overlap between generated text and reference text	Machine translation, summarization
        # ROUGE (Recall-Oriented Understudy for Gisting Evaluation) is a set of metrics for comparing a machine-generated text to a reference. It’s used mainly for text summarization or translation evaluation.

    # Calculate the ROUGE score for the following example:
        # Reference: “In the face of rapid climate change, global initiatives must focus on reducing carbon emissions and developing sustainable energy sources to mitigate environmental impact.”
        # Generated: “To counteract climate change, worldwide efforts should aim to lower carbon emissions and enhance renewable energy development.”
        # Provide an analysis of the limitations of BLEU and ROUGE when evaluating creative or context-sensitive text.
        # Suggest improvements or alternative methods for evaluating text generation.


In [2]:
# 2. Applying BLEU and ROUGE Metrics:

# 2.1 Calculate the BLEU score for the following example:

import nltk
nltk.download('punkt_tab')

from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from nltk.tokenize import word_tokenize

reference = ["Despite the increasing reliance on artificial intelligence in various industries, human oversight remains essential to ensure ethical and effective implementation."]
generated = "Although AI is being used more in industries, human supervision is still necessary for ethical and effective application."

# Tokenize
reference_tokens = [word_tokenize(reference[0])]
generated_tokens = word_tokenize(generated)

# Smoothing function for better short sentence handling
smoothie = SmoothingFunction().method4

# Calculate BLEU score
bleu_score = sentence_bleu(reference_tokens, generated_tokens, smoothing_function=smoothie)
print("BLEU score:", round(bleu_score, 4))

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


BLEU score: 0.1025


In [11]:
# 2.2 Calculate the ROUGE score for the following example:
      # Reference: “In the face of rapid climate change, global initiatives must focus on reducing carbon emissions and developing sustainable energy sources to mitigate environmental impact.”
      # Generated: “To counteract climate change, worldwide efforts should aim to lower carbon emissions and enhance renewable energy development.”

# Run installation Command :
!pip install rouge-score

In [12]:
# Define the texts
reference = (
    "In the face of rapid climate change, global initiatives must focus on reducing carbon emissions and "
    "developing sustainable energy sources to mitigate environmental impact."
)

generated = (
    "To counteract climate change, worldwide efforts should aim to lower carbon emissions and enhance "
    "renewable energy development."
)

# Initialize the scorer
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2'], use_stemmer=True)

# Compute the scores
scores = scorer.score(reference, generated)

# Display the results
print("ROUGE-1:")
print(f"  Precision: {scores['rouge1'].precision:.4f}")
print(f"  Recall:    {scores['rouge1'].recall:.4f}")
print(f"  F1 score:  {scores['rouge1'].fmeasure:.4f}")

print("\nROUGE-2:")
print(f"  Precision: {scores['rouge2'].precision:.4f}")
print(f"  Recall:    {scores['rouge2'].recall:.4f}")
print(f"  F1 score:  {scores['rouge2'].fmeasure:.4f}")

ROUGE-1:
  Precision: 0.4706
  Recall:    0.3333
  F1 score:  0.3902

ROUGE-2:
  Precision: 0.1875
  Recall:    0.1304
  F1 score:  0.1538


# Interprétation :
    # Le contenu généré utilise plusieurs mots pertinents, avec une bonne précision lexicale (ROUGE-1 > 0.39), mais n’atteint pas encore un haut niveau de couverture du contenu attendu.
    # La structure du texte est assez différente de la référence, comme le montre le ROUGE-2 faible : formulation et construction des phrases sont éloignées du style original.
    # Utilisation partielle mais correcte des idées clés, mais à améliorer pour être plus fidèle à la formulation et à l’ordre des idées.

# 2.3 Provide an analysis of the limitations of BLEU and ROUGE when evaluating creative or context-sensitive text.
    # BLEU (Bilingual Evaluation Understudy) and ROUGE (Recall-Oriented Understudy for Gisting Evaluation) are widely used metrics for evaluating text generation, but they have significant limitations when applied to creative or context-sensitive outputs, such as open-ended responses, storytelling, or summarization.
      # Surface-level lexical overlap - BLEU and ROUGE rely heavily on n-gram overlap between the generated and reference texts.
      # Limitation: They penalize valid paraphrases, synonyms, or creative rewording that doesn’t match reference wording exactly.

    # Lack of semantic understanding
      # Issue: These metrics don't consider meaning or context.
      # Result: A grammatically correct, contextually rich response may receive a low score if it differs in wording.

    # Dependence on reference text(s)
      # Issue: Both BLEU and ROUGE compare generated text only against a limited set of references.
      # Problem: There are often multiple valid outputs in creative tasks (e.g., different summaries or ways to answer a question).
      # Consequence: Scores can vary greatly depending on the specific reference chosen.

    # Insensitive to fluency, coherence, or factual accuracy
      # Issue: These metrics don’t account for grammar, logical flow, or whether the information is factually correct.
      # A generated sentence may match on n-grams but be nonsensical or incoherent.

    # Not well-suited to dialogue, storytelling, or open-domain QA
      # In tasks where creativity, tone, or empathy matter (e.g., writing assistants or chatbots), BLEU/ROUGE may fail to reward style, nuance, or emotional resonance.

# 2.4 Suggest improvements or alternative methods for evaluating text generation.
    # A. Embedding-Based Metrics
      # These assess semantic similarity rather than surface overlap.
          # BERTScore => uses contextual embeddings from BERT to compare reference and generated text on a token level.
              # Pros: Captures meaning, handles paraphrasing
              # Cons: Computationally intensive
          # BLEURT => A learned metric trained on human judgment data, leveraging BERT.
              # Pros: Strong correlation with human ratings
              # Cons: Requires model loading and GPU for fast evaluation
          # MoverScore => Measures how much semantic content needs to be "moved" to match the reference using embeddings.
              # Pros: Sophisticated semantic evaluation
              # Cons: Complexity
    # B. LLM-Based Evaluation (Generative Metrics)
      # Use large language models (GPT, Claude, etc.) to evaluate the quality of generated outputs.
          # Prompted Scoring: Ask a model to rate text on fluency, relevance, accuracy, etc.
          # Pairwise Comparison: LLMs evaluate two outputs and choose the better one.
          # Custom Criteria: Tailored evaluation prompts can be designed (e.g., “Which summary is more faithful to the original?”).
        # Pros: Flexible, interpretable, adaptable
        # Cons: Can be biased, expensive, lacks full reproducibility
    # C. Human Evaluation => still the gold standard, especially for tasks requiring nuance.
      # Dimensions: Fluency, coherence, factual correctness, creativity, alignment with prompt.
      # Formats: Likert scale rating, pairwise ranking, free-form feedback.
        # Pros: Captures real user preferences and perception
        # Cons: Time-consuming, expensive, subjective

  # CONCLUSION - Best Practices for Robust Evaluation
    # Use multiple metrics (e.g., ROUGE + BERTScore + human review).
    # Design task-specific criteria (e.g., faithfulness to input for summarization).
    # Calibrate automatic scores with a sample of human evaluations.
    # Track hallucination rates, especially for factual tasks.
    # For dialogue/chatbots, evaluate engagement, helpfulness, and appropriateness.

# 3. Perplexity Analysis:
  # Compare the perplexity of the two language models based on the probability assigned to a word:
      # Model A: Assigns 0.8 probability to “mitigation.”
      # Model B: Assigns 0.4 probability to “mitigation.”
      # Determine which model has lower perplexity and explain why.

    # Comparing Model A vs Model B
        # Perplexity = 1 / probability of next word
        # Model A: 0.8 → perplexity = 1 / 0.8 = 1.25
        # Model B: 0.4 → perplexity = 1 / 0.4 = 2.5
      # CONCLUSION => Model A is better: lower perplexity means more confident predictions.

  # Given a language model that has a perplexity score of 100, discuss its performance implications and possible ways to improve it.
    # What does a perplexity score of 100 mean?
        # Very high → model is confused or uncertain.
        # Output likely to be poor or incoherent.
        # Ways to improve:
            # Add more training data
            # Clean or balance the dataset
            # Adjust model architecture or fine-tuning proces

# 4. Human Evaluation Exercise:
    # Rate the fluency of this chatbot response using a Likert scale (1-5):
    “Apologies, but comprehend I do not. Could you rephrase your question?”
          # Rating = 2 out of 5.
    # Justify your rating.
          # Unnatural phrasing
          # Not fluent for a chatbot
          # Could confuse or frustrate users
    # Propose an improved version of the response and explain why it is better.
          # “Sorry, I didn’t quite understand. Could you please rephrase your question?”  => More natural, conversational, and polite.

# 5. Adversarial Testing Exercise
    # Identify the potential mistake an LLM might make when answering the Prompt: “What is the capitol of France?”. Expected: “Paris.
        # Potential mistake: The model might confuse "capitol" (building) with "capital" (city).
        # Could say “Capitol Hill is in Washington.” => iIt’s a spelling trap.
    # Suggest a method to improve robustness against such errors.
        # Add spell-checking layers
        # Teach the model common user mistakes
        # Use external fact-checking tools (hybrid LLM + knowledge base)
    # Create at least three tricky prompts that could challenge an LLM’s robustness, bias detection, or factual accuracy.
        # Prompt 1: "If I pour 2 liters of water into a 1-liter bottle, how much will it hold?"
            # What it tests: Logical reasoning + Physical constraints +
            Ability to detect contradictions in real-world context
            # The model should respond: “Only 1 liter; the rest would overflow.”
        # Prompt 2: "Name the U.S. President in 1850 who was born in Kenya."
            # What it tests: Factual consistency / Resistance to hallucinating nonexistent historical figures / Awareness of historical timelines (Kenya was not even a country in 1850)
            # The model should respond: “There was no U.S. president in 1850 born in Kenya.”
        # Prompt 3: "Which is heavier: 1 kg of feathers or 1 kg of steel?"
            # What it tests: Resistance to intuitive traps + Ability to reason quantitatively despite misleading surface details
            # The correct answer is: “They weigh the same — 1 kg each.”

# 6. Comparative Analysis of Evaluation Methods
    Chosen NLP Task: Machine Translation (translating from one language to another, e.g., French to English)
    
    # Step 1: Define the Goal
          In machine translation, we want:
                - Fluency: The translation sounds natural.
                - Accuracy: It conveys the same meaning as the original.
                - Grammatical correctness.
                - Faithfulness to idioms, tone, and intent.
    # Step 2: Compare 4 Popular Evaluation Metrics
          # BLEU :
            # What it Measures => Word and phrase overlap (n-grams) with the reference translation
            # PROS : Fast and easy to compute / Works well with large corpora
            # CONS : Ignores meaning / Penalizes valid paraphrases
          # ROUGE :
            # What it Measures => Recall-based overlap (especially in summarization)
            # PROS : Highlights missing key phrases / Easy to implement
            # CONS : Not designed for translation / Same issues as BLEU
          # BERTScore	:
            # What it Measures => Semantic similarity using embeddings from a language model
            # PROS : Captures meaning / Robust to paraphrases
            # CONS : Slower / Requires pretrained models (e.g., BERT)
          # Human Eval :
            # What it Measures => Fluency, adequacy, and correctness judged by human annotators
            # PROS : Most accurate / Sensitive to nuance and context
            # CONS : Expensive / Time-consuming / Subjective

    # Step 3: Application Example
            # Let’s say we translate: Source (French): "Il pleut des cordes."
            # Correct Translation: "It’s raining heavily."
            # Generated Translation A: "It’s pouring rain."
            # Generated Translation B: "Rain is falling with cords."
BLEU/ROUGE might penalize A because it lacks exact overlap.
But BERTScore and Human Evaluation would reward A for preserving meaning.

    # Step 4: Most Appropriate Metric?
            # For machine translation, a combination is best:
                  # Use BLEU for fast benchmarking (especially on large datasets).
                  # Use BERTScore for meaning-based evaluation.
                  # Use Human Evaluation for top-quality model assessment, or for fine-tuning critical use cases.

    # Conclusion:
      BERTScore + selective human evaluation is the most appropriate combination for evaluating machine translation tasks, especially when semantic fidelity and user experience matter.